pocs/camera/zwo.py

```diff
@@ -0,0 +1,446 @@
+        timer = CountdownTimer(duration=self._timeout)
+        try:
+            exposure_status = Camera._ASIDriver.get_exposure_status(self._camera_ID)
+            while exposure_status == 'WORKING':
+                if timer.expired():
+                    msg = "Timeout waiting for exposure on {} to complete".format(self)
+                    raise error.Timeout(msg)
+                time.sleep(0.01)
+                exposure_status = Camera._ASIDriver.get_exposure_status(self._camera_ID)
+        except RuntimeError as err:
+            # Error returned by driver at some point while polling
+            self.logger.error('Error while waiting for exposure on {}: {}'.format(self, err))
+            raise err
+        else:
+            if exposure_status == 'SUCCESS':
+                try:
+                    image_data = Camera._ASIDriver.get_exposure_data(self._camera_ID,
+                                                                     width,
+                                                                     height,
+                                                                     self.image_type)
+                except RuntimeError as err:
+                    raise error.PanError('Error getting image data from {}: {}'.format(self, err))
+                else:
+                    fits_utils.write_fits(image_data, header, filename, self.logger, exposure_event)
+            elif exposure_status == 'FAILED':
+                raise error.PanError("Exposure failed on {}".format(self))
+            elif exposure_status == 'IDLE':
+                raise error.PanError("Exposure missing on {}".format(self))
+            else:
+                raise error.PanError("Unepected exposure status on {}: '{}'".format(
+                    self, exposure_status))
+        finally:
+            exposure_event.set()  # write_fits will have already set this, *if* it got called.
+            self._exposure_lock.release()
+
+    def _fits_header(self, seconds, dark):
+        header = super()._fits_header(seconds, dark)
+        header.set('CAM-GAIN', self.gain, 'Internal units')
+        header.set('CAM-BITS', int(get_quantity_value(self._info['bit_depth'], u.bit)),
+                   'ADC bit depth')
+        header.set('XPIXSZ', get_quantity_value(self._info['pixel_size'], u.um), 'Microns')
+        header.set('YPIXSZ', get_quantity_value(self._info['pixel_size'], u.um), 'Microns')
+        header.set('EGAIN', get_quantity_value(self.egain, u.electron / u.adu), 'Electrons/ADU')
+        return header
+
+    def _refresh_info(self):
+        self._info = Camera._ASIDriver.get_camera_property(self._camera_index)
+
+    def _control_getter(self, control_type):
+        if control_type in self._control_info.keys():
+            return Camera._ASIDriver.get_control_value(self._camera_ID, control_type)
+        else:
+            raise NotImplementedError("'{}' has no '{}' parameter".format(self.model, control_type))
```

Comment:
Maybe an `AttributeError` here instead?

Reply:
I used `NotImplementedError` here (even though this isn't really what that's for) because there is already a lot of code (e.g. large swathes of `pocs/tests/test_camera.py`) that catches `NotImplementedError`s from the `Camera` property getters and setters as a way of determining what capabilities the `Camera` has. Maybe I should use a new custom exception (e.g. NotSupportedError) that inherits from NotImplementedError (and PanError)?

Reply:
> Maybe I should use a new custom exception (e.g. `NotSupportedError`) that inherits from `NotImplementedError` (and `PanError`)?

If not too much work, I do like this idea.

Reply:
OK, added a couple of simple custom exceptions. I should use `PanError` based exceptions more. There are a lot of places in my code where I use builtin exceptions and have to log the error separately.